Verify cover supported features per window covering type

jeverley opened the pull request:
## What's changed
- The `current_position` and `current_tilt_position` state attributes now report `None` if their features are not supported by the device, resolves [#142352](https://github.com/home-assistant/core/issues/142352)
  - This is required because some devices are reporting a position of 0 rather than null
- Revise `supported_features` logic so it _excludes_ tilt only window covering types when enabling lift features
  - This means that lift features will still be enabled as a fallback if the `window_covering_type` attribute value is undefined by the device
  - Previously lift features were enabled for all devices, they should not be enabled for `WCT.Shutter` or `WCT.Tilt_blind_tilt_only` types, see _Table 7-41. Window Covering Type_ from the [Cluster Library Specification 07-5123 Revision 8](https://zigbeealliance.org/wp-content/uploads/2021/10/07-5123-08-Zigbee-Cluster-Library.pdf)
<img alt="image" width="544" src="https://private-user-images.githubusercontent.com/32534428/431515604-4d9c6bd9-ab1f-4ed5-8a98-3ef301af61f0.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NDQxODYxMTYsIm5iZiI6MTc0NDE4NTgxNiwicGF0aCI6Ii8zMjUzNDQyOC80MzE1MTU2MDQtNGQ5YzZiZDktYWIxZi00ZWQ1LThhOTgtM2VmMzAxYWY2MWYwLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTA0MDklMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwNDA5VDA4MDMzNlomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWRhZDNjYTZmNGQ4OWZiNTg4NjVhYmM4MTAwMWExZjBhZmYxYWZiNTQ1OWEwZjE2Y2ExMjcyYzUzNzgxNDUxMGMmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.I4rHEKyuJZkZfnpzq6inZybVxfmttandBPvTje_oRBU">

- Updates to cover platform tests
  - Check lift only devices don't have tilt features enabled and that state attribute `current_tilt_position` is always None
  - Check tilt only devices don't have lift features enabled and that state attribute `current_position` is always None

## Fixes issue
- Some devices report 0 (open) rather than null values for non-functional tilt/lift position attributes
https://github.com/home-assistant/core/issues/142352


## Related
- https://github.com/zigpy/zha/pull/376


FYI @TheJulianJES, @puddly, @dmulcahey

jeverley committed:
Amend _attr_supported_features logic and position attributes to respect window_covering_type

jeverley committed:
Check `supported_features` rather than `_attr_supported_features` in current_position properties

jeverley committed:
Test for supported features and attributes on lift/tilt only devices

Apply pre-commit auto fixes

jeverley committed:
Use same name for _supported_features as other platform handlers

jeverley committed:
Compute supported features prior to initialising positional logic

jeverley committed:
Move timeout constants into Cover class as they aren't used by Shade

- Also use them in the tests rather than defining them separately

pre-commit-ci[bot] committed:
Apply pre-commit auto fixes

puddly commented on zha/application/platforms/cover/__init__.py:
```diff
@@ -669,7 +676,7 @@ def state(self) -> dict[str, Any]:
     @functools.cached_property
     def supported_features(self) -> CoverEntityFeature:
         """Return supported features."""
-        return self._attr_supported_features
```
For consistency with the other platforms and attributes, let's keep the `_attr` prefix for now. These are meant to mirror those in Home Assistant: https://github.com/home-assistant/core/blob/3aae280de50a92cdfcf5c04b088fe9178fba0b2e/homeassistant/components/cover/__init__.py#L240

jeverley replied:
Hey @puddly, I've reverted the naming to include the `_attr` prefix, I've also adjusted the `recompute_capabilities` method so that it only updates the final value after evaluating all criteria.

puddly commented on zha/application/platforms/cover/__init__.py:
```diff
@@ -106,32 +104,37 @@ def __init__(
+                | CoverEntityFeature.CLOSE_TILT
+                | CoverEntityFeature.STOP_TILT
+                | CoverEntityFeature.SET_TILT_POSITION
+            )
```
Just for future reference, this is the ZCL spec table that explains the tilt and lift support of each window covering type:

<img width="544" alt="image" src="https://github.com/user-attachments/assets/4d9c6bd9-ab1f-4ed5-8a98-3ef301af61f0" />

jeverley replied:
I've now added this to the PR description for better visibility.

jeverley committed:
Use `_attr_supported_features` name to mirror core cover component

Addresses feedback https://github.com/zigpy/zha/pull/410#discussion_r2033768660

puddly committed:
Merge branch 'dev' into cover-handle-lift-only-devices-reporting-tilt-position

TheJulianJES commented on zha/application/platforms/cover/__init__.py:
```diff
@@ -123,6 +116,10 @@ async def async_stop_cover(self, **kwargs: Any) -> None:
 class Cover(BaseCover):
     """Representation of a ZHA cover."""
 
+    DEFAULT_MOVEMENT_TIMEOUT: float = 5
+    LIFT_MOVEMENT_TIMEOUT_RANGE: float = 300
+    TILT_MOVEMENT_TIMEOUT_RANGE: float = 30
```
This doesn't really need to be in the class, right?
Seems a bit weird to have constants (in caps) as class attributes.

jeverley replied:
Fair enough, thanks for the amendments @puddly

puddly committed:
Move constants out of class and into globals

puddly committed:
Update constants in unit tests

puddly committed:
Merge branch 'dev' into cover-handle-lift-only-devices-reporting-tilt-position

puddly committed:
Re-add constant comments

puddly committed:
Oops, fix comment order
